### PR TITLE
Inject DateTimeProvider into PeerConnector

### LINKED ISSUE #877: [PeerConnector] Not using IDateTimeProvider and ILoggerFactory
```
                    this.peerAddressManager.PeerAttempted(peerAddress.Endpoint, DateTimeProvider.Default.GetUtcNow());
```

This is a mistake. The sole purpose of using DateTimeProvider is the possibility to inject custom time source. This code is no better than using DateTime.UtcNow. 

another mistake is that initialization routine is taking `ILogger logger` instead of `ILoggerFactory`. The logger should never be shared and each component should construct its own logger from factory.

### PR DESCRIPTION
**Pull request by fassadlr**
This fixes #877.

### CODE REVIEW

File: src/Stratis.Bitcoin.IntegrationTests/P2P/NodeConnectionTests.cs [hunk 1]
```diff
@@ -40,6 +40,7 @@ public void CanDiscoverAndConnectToPeersOnTheNetwork()
 
             var peerConnectorDiscovery = new PeerConnectorDiscovery(
                 new AsyncLoopFactory(loggerFactory),
+                new DateTimeProvider(),
```
Comment (Aprogiena) [inline src/Stratis.Bitcoin.IntegrationTests/P2P/NodeConnectionTests.cs]: use DateTimeProvider.Default instead

File: src/Stratis.Bitcoin.Tests/P2P/PeerConnectorTests.cs [hunk 2]
```diff
@@ -27,7 +28,7 @@ public void PeerConnectorAddNode_FindPeerToConnectTo_Returns_AddNodePeers()
                 ConnectionManager = new Configuration.Settings.ConnectionManagerSettings()
             };
             nodeSettings.ConnectionManager.AddNode.Add(networkAddressAddNode.Endpoint);
-            var connector = new PeerConnectorAddNode(nodeSettings, peerAddressManager);
+            var connector = new PeerConnectorAddNode(new DateTimeProvider(), nodeSettings, peerAddressManager);
```
Comment (Aprogiena) [inline src/Stratis.Bitcoin.Tests/P2P/PeerConnectorTests.cs]: everywhere

File: src/Stratis.Bitcoin.IntegrationTests/WalletTests.cs [hunk 3]
```diff
@@ -195,7 +195,7 @@ public void WalletCanReorg()
                 TestHelper.WaitLoop(() => stratisReceiver.FullNode.WalletManager().GetSpendableTransactionsInWallet("mywallet").Any());
                 var newamount = stratisReceiver.FullNode.WalletManager().GetSpendableTransactionsInWallet("mywallet").Sum(s => s.Transaction.Amount);
                 Assert.Equal(Money.COIN * 110, newamount);
-                Assert.True(stratisReceiver.FullNode.WalletManager().GetSpendableTransactionsInWallet("mywallet").Any(b => b.Transaction.BlockHeight == null));
+                Assert.Contains(stratisReceiver.FullNode.WalletManager().GetSpendableTransactionsInWallet("mywallet"), b => b.Transaction.BlockHeight == null);
```
Comment (Aprogiena) [inline src/Stratis.Bitcoin.IntegrationTests/WalletTests.cs]: little offtopic in this PR i suppose, but hopefully ok

File: src/Stratis.Bitcoin/P2P/PeerAddressManager.cs [hunk 4]
```diff
@@ -94,23 +94,28 @@ public interface IPeerAddressManager : IDisposable
...
+        /// <inheritdoc />
+        public ConcurrentDictionary<IPEndPoint, PeerAddress> Peers { get; private set; }
+
+        /// <summary>The file name of the peers file.</summary>
+        internal const string PeerFileName = "peers.json";
+
+        /// <inheritdoc />
+        public DataFolder PeerFilePath { get; set; }
+
+        /// <summary>Constructor used by unit tests.</summary>
```
Comment (Aprogiena) [inline src/Stratis.Bitcoin/P2P/PeerAddressManager.cs]: in general such constructors seem like hacks, is there a plan to remove them in future?

File: src/Stratis.Bitcoin/P2P/PeerAddressManager.cs [hunk 5]
```diff
@@ -94,23 +94,28 @@ public interface IPeerAddressManager : IDisposable
...
         public PeerAddressManager()
         {
             this.Peers = new ConcurrentDictionary<IPEndPoint, PeerAddress>();
         }
 
-        public PeerAddressManager(DataFolder peerFilePath) : this()
+        /// <summary>Constructor used by dependency injection.</summary>
```
Comment (Aprogiena) [inline src/Stratis.Bitcoin/P2P/PeerAddressManager.cs]: missing `<param>`

File: src/Stratis.Bitcoin/P2P/PeerAddressManagerBehaviour.cs [hunk 6]
```diff
@@ -17,6 +17,24 @@ namespace Stratis.Bitcoin.P2P
...
+        /// <summary>Provider of time functions.</summary>
+        private readonly IDateTimeProvider dateTimeProvider;
+
+        /// <summary>
+        /// See <see cref="PeerAddressManagerBehaviourMode"/> for the different modes and their
+        /// explanations.
+        /// </summary>
+        public PeerAddressManagerBehaviourMode Mode { get; set; }
+
+        /// <summary>Peer address manager instance, see <see cref="IPeerAddressManager"/>.</summary>
```
Comment (Aprogiena) [inline src/Stratis.Bitcoin/P2P/PeerAddressManagerBehaviour.cs]: you don't really need to tell the reader to see the type of the thing you are commenting. see should be used when you are talking about something else, ... same above with PeerAddressManagerBehaviourMode 

also these comments are not very useful in general
i can see those a lot in your code, not really happy about it :(

File: src/Stratis.Bitcoin/P2P/PeerConnectorConnect.cs [hunk 7]
```diff
@@ -18,20 +18,21 @@ public sealed class PeerConnectorConnectNode : PeerConnector
...
         public PeerConnectorConnectNode(
             IAsyncLoopFactory asyncLoopFactory,
+            IDateTimeProvider dateTimeProvider,
             ILoggerFactory loggerFactory,
             Network network,
             INetworkPeerFactory networkPeerFactory,
             INodeLifetime nodeLifetime,
             NodeSettings nodeSettings,
             IPeerAddressManager peerAddressManager)
             :
```
Comment (Aprogiena) [inline src/Stratis.Bitcoin/P2P/PeerConnectorConnect.cs]: does the colon deserve its own line?

File: src/Stratis.Bitcoin.IntegrationTests/P2P/NodeConnectionTests.cs (hunk 1 above)
Comment (fassadlr) [reply src/Stratis.Bitcoin.IntegrationTests/P2P/NodeConnectionTests.cs]: Fixed

File: src/Stratis.Bitcoin.Tests/P2P/PeerConnectorTests.cs (hunk 2 above)
Comment (fassadlr) [reply src/Stratis.Bitcoin.Tests/P2P/PeerConnectorTests.cs]: OK fixed this in all the applicable tests.

File: src/Stratis.Bitcoin/P2P/PeerAddressManager.cs (hunk 4 above)
Comment (fassadlr) [reply src/Stratis.Bitcoin/P2P/PeerAddressManager.cs]: I normally don't see an issue with having code that facilitates unit testing... But if you insist I can remove this in this instance. I don't mind either way.

File: src/Stratis.Bitcoin/P2P/PeerAddressManager.cs (hunk 5 above)
Comment (fassadlr) [reply src/Stratis.Bitcoin/P2P/PeerAddressManager.cs]: Fixed

File: src/Stratis.Bitcoin/P2P/PeerAddressManagerBehaviour.cs (hunk 6 above)
Comment (fassadlr) [reply src/Stratis.Bitcoin/P2P/PeerAddressManagerBehaviour.cs]: I don't think when IPeerAddressManager is properly documented, that I need to rewrite all the comments? That just seems a bit superfluous.

File: src/Stratis.Bitcoin/P2P/PeerConnectorConnect.cs (hunk 7 above)
Comment (fassadlr) [reply src/Stratis.Bitcoin/P2P/PeerConnectorConnect.cs]: OK I've moved it next to the constructor line.

File: src/Stratis.Bitcoin/P2P/PeerAddressManager.cs (hunk 4 above)
Comment (Aprogiena) [reply src/Stratis.Bitcoin/P2P/PeerAddressManager.cs]: i'm not sure, i have a feeling that if you need special constructors for that, something may be wrong with how the thing is designed, but i haven't put a deeper thought into it

Comment (dangershony) [reply src/Stratis.Bitcoin/P2P/PeerAddressManager.cs]: It is wrong yes, if this is temporary (like I did in node where we did not have an interface) then ok.
But normally we need to inject the dependencies and use mock interfaces.

Comment (fassadlr) [reply src/Stratis.Bitcoin/P2P/PeerAddressManager.cs]: OK guys, I'll remove this constructor 👍

Comment (fassadlr) [reply src/Stratis.Bitcoin/P2P/PeerAddressManager.cs]: And the ones in the subclasses.

**Pull request merged**